File: classes/201-f25/Comparing-Python-and-Kotlin/linereader.py
"""
   linereader.py
   Jeff Ondich, 2013-01-04
   Anna Rafferty

   Modified by Tanya Amert for Spring 2025

   Reads lines from a file, prints them in upper-case, and prints
   a count of the lines at the end.
   
   Intended as the Python half of parallel examples in Python and
   Java. See LineReader.java.
"""

import sys

# Try to open the file
inputFilePath = "somelines.txt"
try:
    inputFile = open(inputFilePath)
except Exception as e:
    print("Cannot open", inputFilePath, file=sys.stderr)
    exit()

# Do the actual reading/printing
numberOfLines = 0
for line in inputFile:
    print(line.upper(), end='')
    numberOfLines += 1
print("\nNumber of lines:", numberOfLines)

File: classes/201-f25/Comparing-Python-and-Kotlin/LineReader.kt
/*
 * LineReader.kt
 * Tanya Amert, 2025-03-31
 * Based on LineReader.java by Jeff Ondich, Anna Rafferty
 *
 * A very brief introduction to reading and writing from a file,
 * excessively commented for tutorial purposes.
 *
 * This is the Kotlin half of a pair of parallel examples in Python and Kotlin.
 * See linereader.py.
 *
 * Try a few things:
 *
 * 1. Look at the small text file called somelines.txt and then compile and run LineReader.
 *    ("kotlinc LineReader.kt" followed by "kotlin LineReaderKt.class")
 *
 * 2. What happens when you change the inputFilePath variable to "nonexistent.txt"?
 *
 * 3. What happens when you remove the try/catch blocks and just do
 *    "inputFile.forEachLine() { ... }"?
 *
 * 4. Visit the Kotlin String documentation and play around with some String
 *    methods other than uppercase.
 */

import java.io.File // it's secretly Java under the hood
import java.io.FileNotFoundException
import kotlin.system.exitProcess

fun main() {
    // Create a File object connected to the file you want to read
    val inputFilePath: String = "somelines.txt"
    val inputFile: File = File(inputFilePath)

    // Actually reading from the file can result in an exception, so
    // that's why there is a try/catch block to catch the exception.
    // It allows the programmer to build in a way for the program to
    // recover from the error (in this case, by just printing the error
    // for the user).
    var numberOfLines: Int = 0
    try {
        // Get one line at a time from the file, and print each line in upper
        // case to standard output
        inputFile.forEachLine() {
            println(it.uppercase()) // it refers to each element of the sequence from forEachLine()
            numberOfLines++
        }
    } catch (e: FileNotFoundException) {
        println(e)     // to stdout not stderr, but close enough
        exitProcess(1) // anything nonzero is an error status code
    }

    println("Number of lines: $numberOfLines")
}